File: ui/src/api/hooks/plannedTransactions.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import {
  api,
  type NewPlannedTransaction,
  type PlannedTransaction,
  type PlannedTransactionPatch,
  type PlannedTxnFilter,
} from "../openapiClient";
import { unwrap } from "../openapiClient";
import { isBackendAvailable } from "../../utils/runtime";

export function usePlannedTransactions(filter: Partial<PlannedTxnFilter> = {}) {
  return useQuery<PlannedTransaction[]>({
    queryKey: ["planned-transactions", filter],
    queryFn: async () => {
      return unwrap(api.listPlannedTransactions({
        status: filter.status ?? null,
        dueBefore: filter.dueBefore ?? null,
      }));
    },
    enabled: isBackendAvailable(),
  });
}

export function useCreatePlannedTransaction() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (input: NewPlannedTransaction) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      return unwrap(api.createPlannedTransaction(input));
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["planned-transactions"] });
      qc.invalidateQueries({ queryKey: ["journey-status"] });
    },
  });
}

export function useUpdatePlannedTransaction() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ id, patch }: { id: string; patch: PlannedTransactionPatch }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      return unwrap(api.updatePlannedTransaction(id, patch));
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["planned-transactions"] });
      qc.invalidateQueries({ queryKey: ["journey-status"] });
    },
  });
}

export function useDeletePlannedTransaction() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: string) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      await unwrap(api.deletePlannedTransaction(id));
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["planned-transactions"] });
      qc.invalidateQueries({ queryKey: ["journey-status"] });
    },
  });
}

File: crates/finsight-bindings/src/commands/planned_transactions.rs
use crate::error::AppResult;
use crate::AppState;
use finsight_core::models::{
    NewPlannedTransaction, PlannedTransaction, PlannedTransactionPatch, PlannedTxnFilter,
};

#[tauri::command]
#[specta::specta]
pub async fn list_planned_transactions(
    state: tauri::State<'_, AppState>,
    filter: PlannedTxnFilter,
) -> AppResult<Vec<PlannedTransaction>> {
    finsight_api::commands::planned_transactions::list_planned_transactions(&state.api, filter)
        .await
}

#[tauri::command]
#[specta::specta]
pub async fn get_planned_transaction(
    state: tauri::State<'_, AppState>,
    id: String,
) -> AppResult<Option<PlannedTransaction>> {
    finsight_api::commands::planned_transactions::get_planned_transaction(&state.api, id).await
}

#[tauri::command]
#[specta::specta]
pub async fn create_planned_transaction(
    state: tauri::State<'_, AppState>,
    input: NewPlannedTransaction,
) -> AppResult<PlannedTransaction> {
    finsight_api::commands::planned_transactions::create_planned_transaction(&state.api, input)
        .await
}

#[tauri::command]
#[specta::specta]
pub async fn update_planned_transaction(
    state: tauri::State<'_, AppState>,
    id: String,
    patch: PlannedTransactionPatch,
) -> AppResult<PlannedTransaction> {
    finsight_api::commands::planned_transactions::update_planned_transaction(
        &state.api, id, patch,
    )
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn delete_planned_transaction(
    state: tauri::State<'_, AppState>,
    id: String,
) -> AppResult<()> {
    finsight_api::commands::planned_transactions::delete_planned_transaction(&state.api, id).await
}
